### ISSUE TITLE
Cloud repo source import fails #736 (Still a problem)

### ISSUE THREAD

**Issue by ryust** [GoogleCloudPlatform/google-cloud-intellij#1174]
As the title indicates, I am having the EXACT same problem as described in issue #736. Either the problem was not fully resolved or it has come back. I have successfully imported Android Studio projects into different Google Cloud repositories in the past, but now I cannot on two new projects I have.

**Comment by patflynn** [comment]
@ryust What error message are you seeing?  

CSR is moving away from having a 'default' repository available at initialization.  Our tooling still expects it and we're working on fixing that. https://github.com/GoogleCloudPlatform/google-cloud-intellij/issues/1024

**Comment by ryust** [comment]
The exact same errors as show in the image for #736. Everything is the same problem I am encountering. There is already a 'default' repository in Google Cloud and it looks just like the others I have successfully imported to with other Android Studio projects in the past.

**Comment by ryust** [comment]
![error](https://cloud.githubusercontent.com/assets/289010/21189825/84d9148a-c1e5-11e6-9246-6a72249e7c16.png)

**Comment by etanshaul** [comment]
@ryust just to clarify. Are you getting this error when attempting to push your project to Google Cloud Repositories, or when trying to clone a repo from GCR?

**Comment by ryust** [comment]
It occurs on the first time **push** of the project to Google Cloud. The repository is empty.

**Comment by etanshaul** [comment]
@ryust just to update you - we are investigating this, but as of yet unable to reproduce the issue. I know you mentioned you followed the same steps as in the linked to issue, but if you have any more details you could provide here that would be great.

**Comment by ryust** [comment]
Sorry, I really don't have any other info. It's an Android Studio project that I enabled VCS integration today and then proceeded to Import into Version Control / Upload Project to Google Cloud. Selected the GC project and then the error. Going to the GC developer console I see the project and the default repro, but no source is in it.

**Comment by Xaver2k** [comment]
exact same problem
-creating new project in google cloud console webpage
-in android studio selecting VCS->Import into Version Control->Upload Project to Google Cloud
-select the project in the popup and press "upload"
-at the bottom of android studio i see "Fetching from google remote host..."
-then i get a popup error saying "Failed to Fetch Remote origin"

i also already had a working VCS project synching with a Google Cloud repository.

**Comment by etanshaul** [comment]
thanks @Xaver2k we added this bug to our next milestone

**Comment by ryust** [comment]
Any feedback on resolving this issue?

**Comment by etanshaul** [comment]
@ryust sorry for the delay on this.

I was able to find out more information. I'm guessing you are using Windows?

This stems from a credentials issue. The error dialog we were showing was not very helpful and we'll update this in an upcoming release. As a temporary workaround you can try these steps:

1) Visit https://source.developers.google.com/auth/start?scopes=https://www.googleapis.com/auth/cloud-platform
2) Login to the account you are using for your cloud project
3) On the page that loads, make note of the line that you are instructed to add to your `.netrc` file (noting that you need to replace your GCP email address where it tells you to)
4) Create a netrc file: if you are on Windows, as I suspect, you can create a file called `_netrc` (note the underscore instead of the period) in your home directory (e.g. C:\Users\[username]). And in the file paste the line from step (3)

Now attempt to import your project again.

Let me know if this helps.

**Comment by patflynn** [comment]
@ryust 

To help with debugging: Do you have, or have you ever had, multiple Google accounts logged in through our plugin?

**Comment by ryust** [comment]
YesOn Jan 24, 2017 11:26 AM, Patrick Flynn <notifications@github.com> wrote:@ryust
To help with debugging: Do you have, or have you ever had, multiple Google accounts logged in through our plugin?

**Comment by ryust** [comment]
No, not on Windows.... using an iMac.On Jan 24, 2017 10:34 AM, Etan Shaul <notifications@github.com> wrote:@ryust sorry for the delay on this.
I was able to find out more information. I'm guessing you are using Windows?
This stems from a credentials issue. The error dialog we were showing was not very helpful and we'll update this in an upcoming release. As a temporary workaround you can try these steps:
Visit https://source.developers.google.com/auth/start?scopes=https://www.googleapis.com/auth/cloud-platformLogin to the account you are using for your cloud projectOn the page that loads, make note of the line that you are instructed to add to your .netrc file (noting that you need to replace your GCP email address where it tells you to)Create a netrc file: if you are on Windows, as I suspect, you can create a file called _netrc (note the underscore instead of the period) in your home directory (e.g. C:\Users[username]). And in the file paste the line from step (3)
Now attempt to import your project again.
Let me know if this helps.

**Comment by etanshaul** [comment]
@ryust. do you mind trying out the netrc instructions on your mac (using `.netrc` instead of `_netrc`) to see if that helps? This will give us more hints at the problem.

**Comment by patflynn** [comment]
@ryust @Xaver2k are you guys trying to import into a project that already has a default repo or in a completely new GCP project?

**Comment by patflynn** [comment]
@ryust @Xaver2k  I've actually been able to consistently reproduce this problem in Android Studio on linux but not in IntelliJ.

What I found was that I already had a `machine source.developers.google.com ..` in my ~/.netrc.  When I commented it out repo import started working again.

Can you guys check if you have an equivalent line in .netrc?

**Comment by ryust** [comment]
Completely new wth no previous importOn Jan 24, 2017 1:21 PM, Patrick Flynn <notifications@github.com> wrote:@ryust @Xaver2k are you guys trying to import into a project that already has a default repo or in a completely new GCP project?

**Comment by ryust** [comment]
Sure... will try that tomorrow as I'm out of the office right now.On Jan 24, 2017 1:18 PM, Etan Shaul <notifications@github.com> wrote:@ryust. do you mind trying out the netrc instructions on your mac (using .netrc instead of _netrc) to see if that helps? This will give us more hints at the problem.

**Comment by ryust** [comment]
I did not have a .netrc file anywhere on my iMac so I created one via the instructions. That fixed the problem and I was able to import the two Android Studio projects into the Google Cloud repository!

Thanks!

**Comment by etanshaul** [comment]
@ryust I'm glad those steps unblocked you.

Will keep you updated on any changes we make around this.

**Comment by ryust** [comment]
This problem has come back on a new Android Studio project that I’m trying to upload to Google Cloud repository. Following the instructions again related to the netrc file, but still no luck and the remote origin problem is back. 

Any suggestions?

**Comment by etanshaul** [comment]
@ryust, just to be sure, did you update the netrc file with the new credentials? You could try removing the netrc file, attempting to upload, then recreating it according to the instructions that come up (and finally uploading again) to see if that helps.

**Comment by ryust** [comment]
Steps I took:
Tried to upload using the old netrc file with old credentials from last time when it worked… failed.
Went to web link and got new credentials and added both old and new to netrc… failed.
Removed the old entry in netrc… failed.
Restarted Android Studio… failed.
Restarted my iMac… failed.
Removed the netrc file… failed.
Removed netrc file… failed.
Went to web link and got new credentials and added only new one to netrc… failed.
What else do you want me to try? And why do the credentials change each time I go to the web link?

-Ron

**Comment by etanshaul** [comment]
sorry to make you jump through hoops, but do you mind trying the instructions under this link (set set up the access via the command line):

https://cloud.google.com/source-repositories/docs/quickstart

Specifically, try the instructions under this section:
Add the Cloud Source Repository as a remote

(Note, if you already have a remote named "google" you'll need to name it something else and change the push command accordingly)

Then, if this works, you can try again via the IDE to see if that clears things up. If not, let me know.

**Comment by etanshaul** [comment]
Just to clarify - if the above works, you can subsequently do a normal git pull/push via the IDE git integration if you choose. No need to set up the project on Cloud Source Repos again once its already been set up.

**Comment by ryust** [comment]
Not trying to be obstinate, but I started and then when I saw all the command line steps I stopped. I do not like running things like that when I don’t fully understand what’s happening or how to recover if there is a problem. I may just forget use Google Cloud for my git repositories.

**Comment by etanshaul** [comment]
I understand your frustration. I wasn't intending you to run all the steps
in that link, just those under the "add the cloud source repository as a
remote" and "push to the cloud source repository" headers. There are just 3
steps, one of which is a traditional git push, and another is adding a git
remote. The first one configures the git credential manager to access your
cloud account.

Like I said I'm not sure this will work, just tying to help troubleshoot
your issue.

**Comment by ryust** [comment]
Well, I gave that a try and another failure.  Here is the output from the terminal:

Rons-iMac:geoohlive Ron$ git config credential.helper gcloud.sh
Rons-iMac:geoohlive Ron$ git remote add google https://source.developers.google.com/p/geooh-live-caf28/r/default
Rons-iMac:geoohlive Ron$ git push google master
error: src refspec master does not match any.
error: failed to push some refs to 'https://source.developers.google.com/p/geooh-live-caf28/r/default' <https://source.developers.google.com/p/geooh-live-caf28/r/default'>

**Comment by etanshaul** [comment]
what is the branch you are on? Do a `git branch` from the directory you ran those commands.

the "src refspec master does not match any." error makes me think you don't have a branch called master. If you are on a different branch you could run it this way instead:

`git push google [your-branch-name]`

Also, could you paste in the contents of your git config file just to make sure there isn't anything else funny going on? Its under [project_root]/.git/config

**Comment by ryust** [comment]
I don’t use branches… I’m a very simple user of git. Attached is the config.

**Comment by ryust** [comment]
Guys, is this thing ever going to work? It’s seems a total mess up when trying to use Google Cloud repository from Android Studio.

After the last manual instructions, I was able to do an initial commit, but since then any additional commits return a forbidden error. This stuff can’t be that hard!

**Comment by patflynn** [comment]
Hi Ron.  The team is limited in resources and working on finishing up App
Engine integration at the moment so we don't have the time at the moment to
fix this issue. We'll most likely investigate a robust fix later this
quarter. i.e. May or June.

Sorry for the difficulties you've been facing.

**Comment by nkibler** [comment]
There is nothing for us to do here since it depends on how the development environment was set up, so I'm closing this.

The issue stems from how git4idea selects auth providers. It always adds a `BaseGitAuthDataProvider` to the front of the auth provider list, which AFAICT, tries to pull the credentials from the global Git credential helper, if set. This heavily depends on the client and could be `netrc`, the Git Credential Manager for Windows, the OSX keychain, or any other Git credential helper implementation.

If this data provider succeeds in finding a username/password combination for this URL (which could be configured at the _domain_ level, `source.developers.google.com`), our custom GCP auth provider will never run. This is a problem (if the helper is not the `gcloud` utility) because the password is actually a short-lived generated access token and **not** your Google account password. This means that if you wanted to use an auth provider _other_ than the GCP one, you need to constantly update the auth token since other credential helpers won't do this for you.

**Quick Fix:** Run `git config --global credential.helper ""` to clear your global Git credential helper. Be aware that this could be reset at any time if you use something else that will set this config value.

**A Better Fix:** If you want to keep your global credential helper set as-is, you can follow some steps to only change the credential helper for a specific repository. The steps below assume you have a Git repository that has **not** been uploaded to Google Cloud yet:

1. If you haven't done so already, [install and initialize the Cloud SDK](https://cloud.google.com/sdk/docs/).
1. Open a terminal and `cd` to your repository's top-level directory.
1. Run the following command to update the credential helper of the repo to the `gcloud` utility:
    ```
    git config credential.helper "!gcloud auth git-helper --account=<YOUR_GOOGLE_ACCOUNT_EMAIL> --ignore-unknown $@"
    ```
1. Go back to Android Studio or IntelliJ and upload the source to Google Cloud like normal.

If you already have a repository uploaded to a Google Cloud Source repository but you cannot push changes, you need to re-clone your repo using the `gcloud` utility by running:
```
gcloud beta source repos clone REPO_NAME
```
It's possible that the `git config credential.helper` value was somehow overwritten for your repo, so re-cloning the repo ensures the credential helper is set up properly again.

**Comment by patflynn** [comment]
FYI I just got this error on linux. 

`2/15/18
2:14 PM	Clone failed
				remote error:
				Cloning into 'delete6'...
				
				
				Invalid authentication credentials.
				
				Please generate a new identifier:
				  https://source.developers.google.com/auth/start?scopes=https://www.googleapis.com/auth/cloud-platform
				`
